src/node/rpc/member_frontend.h

```diff
@@ -985,19 +985,28 @@ namespace ccf
           proposal_id,
           ctx.rpc_ctx->get_request_body(),
           constitution.value());
-        pi->put(
-          proposal_id,
-          {caller_identity.member_id,
-           rv.state,
-           {},
-           {},
-           std::nullopt,
-           rv.failure});
-
-        ctx.rpc_ctx->set_response_status(HTTP_STATUS_OK);
+
         ctx.rpc_ctx->set_response_header(
           http::headers::CONTENT_TYPE, http::headervalues::contenttype::JSON);
         ctx.rpc_ctx->set_response_body(nlohmann::json(rv).dump());
+        if (rv.state == ProposalState::FAILED)
+        {
+          // If the proposal failed to apply, we want to discard the tx and not
+          // apply its side-effects to the KV state.
+          ctx.rpc_ctx->set_response_status(HTTP_STATUS_INTERNAL_SERVER_ERROR);
+        }
+        else
+        {
+          pi->put(
+            proposal_id,
+            {caller_identity.member_id,
+             rv.state,
+             {},
+             {},
+             std::nullopt,
+             rv.failure});
+          ctx.rpc_ctx->set_response_status(HTTP_STATUS_OK);
```

Comment:
```suggestion
          ctx.rpc_ctx->set_response_status(HTTP_STATUS_INTERNAL_SERVER_ERROR);
          return;
        }
        else
        {
          pi->put(
            proposal_id,
            {caller_identity.member_id,
             rv.state,
             {},
             {},
             std::nullopt,
             rv.failure});
          ctx.rpc_ctx->set_response_status(HTTP_STATUS_OK);
          return;
```

We've been bitten before by code in endpoint handlers that looks like (/intends to) produce a final result, but doesn't return. Just adding explicit returns here for safety.